parlai/core/torch_agent.py

```diff
@@ -356,11 +366,48 @@ def observe(self, observation):
         return observation
 
     def act(self):
-        """Calls batch_act with the singleton batch"""
+        """Calls batch_act with the singleton batch."""
         return self.batch_act([self.observation])[0]
 
-    def batch_act(self):
-        raise NotImplementedError("Abstract class: user must implement batch_act")
+    def batch_act(self, observations):
+        batch_size = len(observations)
+        # initialize a table of replies with this agent's id
+        batch_reply = [{'id': self.getID()} for _ in range(batch_size)]
+
+        is_training = any(['labels' in obs for obs in observations])
+
+        vec_obs = [self.vectorize(obs) for obs in observations]
+
+        batch = self.batchify(vec_obs)
+
+        if batch.text_vec is None:
+            return batch_reply
+
+        if is_training:
+            output = self.train_step(batch.text_vec, batch.label_vec,
+                                     batch.candidates)
+        else:
+            output = self.eval_step(batch.text_vec, batch.label_vec,
+                                    batch.candidates)
+
+        if isinstance(output, tuple):
+            predictions = output[0]
```

Comment:
Consider tuple unpacking? `predictions, candidate_preds = output`? Also, it's a bit worrying to have the user implement a method with potentially multiple return types.